src/state.rs

```diff
@@ -109,6 +114,47 @@ pub struct ResolvedDns {
     dns_refresh_rate: std::time::Duration,
 }
 
+/// A SingleFlight can make sure that only the first request to really handle DNS resolving task
+/// for specified hostname and other requests in concurrency need to wait for the task completion.
+#[derive(serde::Serialize, Default, Debug)]
+pub struct SingleFlight {
+    // is_first_req_in can record requests number and it can be used to
+    // check the request number, such as the first request.
+    is_first_req_in: bool,
+    // wait_for_notification can help to implement the multiple waiters and a single notifier
+    // and it should be in Arc and protected by RWLock because of the concurrency envs.
+    #[serde(skip_serializing)]
+    wait_for_notification: Arc<RwLock<Notify>>,
+}
+
+impl Clone for SingleFlight {
+    fn clone(&self) -> Self {
+        SingleFlight {
+            is_first_req_in: self.is_first_req_in,
+            wait_for_notification: self.wait_for_notification.clone(),
+        }
+    }
+}
+
+impl SingleFlight {
+    pub fn new() -> Self {
+        SingleFlight {
+            // is_first_req_in is set to true because
+            // there always is the first request
+            is_first_req_in: true,
+            wait_for_notification: Arc::new(RwLock::new(Notify::new())),
```

Comment:
I don't think we need a lock for this.

See the example usage: https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html#examples-5.

Reply:
Yeah, it's unnecessary to use this lock. I removed it.